quantum/process_keycode/process_autocorrect.c

```diff
@@ -0,0 +1,226 @@
+// Copyright 2021 Google LLC
+// Copyright 2021 @filterpaper
+// SPDX-License-Identifier: Apache-2.0
+// Original source: https://getreuer.info/posts/keyboards/autocorrection
+
+#include "process_autocorrect.h"
+#include <string.h>
+#include "keycode_config.h"
+
+#ifndef AUTOCORRECT_DATA_H
+#    define AUTOCORRECT_DATA_H "autocorrect_data.h"
+#endif
+
+#if __has_include(AUTOCORRECT_DATA_H)
+#    include AUTOCORRECT_DATA_H
+#else
+#    pragma message "Warning! Autocorrect is using the default library!"
+#    include "autocorrect_data_default.h"
+#endif
+
+/**
+ * @brief handler for determining if autocorrect should process keypress
+ *
+ * @param keycode Keycode registered by matrix press, per keymap
+ * @param record keyrecord_t structure
+ * @param typo_buffer_size passed along to allow resetting of autocorrect buffer
+ * @param mods allow processing of mod status
+ * @return true Allow autocorection
+ * @return false Stop processing and escape from autocorrect.
+ */
+__attribute__((weak)) bool process_autocorrect_user(uint16_t *keycode, keyrecord_t *record, uint8_t *typo_buffer_size, uint8_t *mods) {
+    // See quantum_keycodes.h for reference on these matched ranges.
+    switch (*keycode) {
+        // Exclude these keycodes from processing.
+        case KC_LSFT:
+        case KC_RSFT:
+        case KC_CAPS:
+        case QK_TO ... QK_ONE_SHOT_LAYER_MAX:
+        case QK_LAYER_TAP_TOGGLE ... QK_LAYER_MOD_MAX:
+        case QK_ONE_SHOT_MOD ... QK_ONE_SHOT_MOD_MAX:
+            return false;
+
+        // Mask for base keycode from shifted keys.
+        case QK_LSFT ... QK_LSFT + 255:
+        case QK_RSFT ... QK_RSFT + 255:
+            if (*keycode >= QK_LSFT && *keycode <= (QK_LSFT + 255)) {
+                *mods |= MOD_LSFT;
+            } else {
+                *mods |= MOD_RSFT;
+            }
+            *keycode &= 0xFF; // Get the basic keycode.
+            return true;
+#ifndef NO_ACTION_TAPPING
+        // Exclude tap-hold keys when they are held down
+        // and mask for base keycode when they are tapped.
+        case QK_LAYER_TAP ... QK_LAYER_TAP_MAX:
+#    ifdef NO_ACTION_LAYER
+            // Exclude Layer Tap, if layers are disabled
+            // but action tapping is still enabled.
+            return false;
+#    endif
+        case QK_MOD_TAP ... QK_MOD_TAP_MAX:
+            if (!record->tap.count) {
```

Comment:
```suggestion
        case QK_MOD_TAP ... QK_MOD_TAP_MAX:
            if (((((*keycode >> 8) & 0xf) & ~MOD_MASK_SHIFT) == false) && !record->tap.count) {
```
Unfortunately this conditional check is still required for mods other than Shift inside mod-taps to properly disable autocorrect in these test cases:

- "`f`" tap `ALT_T(KC_I)` "`tler`" = auto corrected
- "`f`" hold `SFT_T(KC_A)` + `KC_I` "`tler`" =  auto corrected
- "`f`" hold `CTL_T(KC_S)` + `KC_I` "`tler`" =  cancelled